### PR TITLE
9887 - twisted.web.twcgi can now handle url parameters in python 3

### PR DESCRIPTION
**Pull request by ndg63276**
##  Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9887
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

**Review by rodrigc — APPROVED** [review]
This looks OK

**Comment by rodrigc** [comment]
@ndg63276 Do you think you could add a test to `src/twisted/web/test/test_cgi.py` for this?

**Comment by exarkun** [comment]
What does "Approved" and also "Can you add a test" mean in one review?  It seems kind of ambiguous.

File: src/twisted/web/test/test_cgi.py [hunk 1]
```diff
@@ -363,6 +374,44 @@ class CGIScriptTests(unittest.TestCase):
...
+
+
+    def test_urlParameters(self):
+        """
+        If the CGI script is passed URL parameters, dont fall over,
+        as per ticket 9887.
+        """
+        cgiFilename = os.path.abspath(self.mktemp())
+        with open(cgiFilename, 'wt') as cgiFile:
+            cgiFile.write(URL_PARAMETER_CGI)
```
Comment (exarkun) [inline src/twisted/web/test/test_cgi.py]: This file could also just be a sibling of `test_cgi.py` and then it doesn't need to be written out to a temporary file every time the test runs.

File: src/twisted/web/test/test_cgi.py [hunk 2]
```diff
@@ -363,6 +374,44 @@ class CGIScriptTests(unittest.TestCase):
...
+            return self.p.stopListening()
+
+
+    def startServer(self, cgi):
+        root = resource.Resource()
+        cgipath = util.sibpath(__file__, cgi)
+        root.putChild(b"cgi", PythonScript(cgipath))
+        site = server.Site(root)
+        self.p = reactor.listenTCP(0, site)
+        return self.p.getHost().port
```
Comment (exarkun) [inline src/twisted/web/test/test_cgi.py]: This style of testing is, unfortunately, prone to spurious failures.  `listenTCP(0, ...)` will fail in some cases.  Also, the platform TCP/IP stack is necessary for testing the application changes here.  Instead, you can exercise all of the logic by moving bytes around in-memory.

I'm not sure what the current best practices in Twisted are for this kind of testing but I know that there is `twisted.test.iosim` and `twisted.internet.testing` with both offer tools that might be of use.

File: src/twisted/web/test/test_cgi.py [hunk 3]
```diff
@@ -363,6 +372,45 @@ class CGIScriptTests(unittest.TestCase):
     """
     Tests for L{twcgi.CGIScript}.
     """
+    def tearDown(self):
```
Comment (rodrigc) [inline src/twisted/web/test/test_cgi.py]: Thanks for implementing this test.  The `teardown()` and `startServer` methods have been copied from `class CGITests`, which is fine to get this test going.
Can you refactor these two methods into a new mixin class, and reuse the mixin in `CGITests` and `CGIScriptTests`.

Also, as you are finishing this off, can you run the tox command here: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted

just to double check you are addressing any style issues before pushing to GitHub?

File: src/twisted/web/test/test_cgi.py (hunk 1 above)
Comment (ndg63276) [reply src/twisted/web/test/test_cgi.py]: Sorry, I'm new to this kind of testing. This construction is used repeatedly in test_cgi.py.

File: src/twisted/web/test/test_cgi.py (hunk 2 above)
Comment (ndg63276) [reply src/twisted/web/test/test_cgi.py]: This was just copied from CGITests.startServer in the same file.

**Comment by ndg63276** [comment]
```
  lint: commands succeeded
  newsfragment: commands succeeded
  manifest-checker: commands succeeded
  congratulations :)

**Review by rodrigc — APPROVED** [review]
I think this looks OK. 

It looks the the use case you discovered on Python 3 was missed by the existing tests.  Thanks for writing a new test to catch this use case.

The test you wrote follows the convention of the existing CGI tests, so that is OK.

It would be possible to refactor the CGI tests to not actually start a server which listens on a socket as pointed out by exarkun.  If you are motivated enough to dig into the Twisted test code and make that change, please file another PR to make that change.
However that might be a bit intimidating for a new contributor to tackle, so don't feel pressured to do that.

What you contributed in this PR is a valid fix with a nice test to get things going.

**Pull request merged**